### DISCUSSION TITLE
What's the difference between nodal variable and elemental variable in Kernel?

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#22686] [Q&A General] [answered]
I didn't find a page about this question. I remember that computeQpResidual() is executed on each each quadrature point and the node value is calculated by the qp point value. So what is the specific difference between nodal and elemental variable?

**Comment by GiudGiud** [comment, marked as answer]
Hello

The difference between a nodal and an elemental variable is where the degrees of freedoms (dofs) reside. Both elements and nodes are DoFObject in libmesh, so they can "hold" (simplifying) Qp values of a variable. 

A nodal variable may be uniquely defined by its values on nodes (think about a Lagrange variable).
An elemental variable is uniquely defined by its values in various places on each element. This allows for discontinuity from one element to another.

There are mixed nodal - elemental variables that hold dofs on both elements and nodes. 

Guillaume

**Reply by Ethan-xj** [reply]
"A nodal variable may be uniquely defined by its values on nodes (think about a Lagrange variable)."
Does the "nodes" include quadrature point?
If doesn't, why we should add [_qp] behind the variable in computeQpResidual()?

**Reply by GiudGiud** [reply]
> Does the "nodes" include quadrature point? 
This depend on the quadrature. Typically not
> If doesn't, why we should add [_qp] behind the variable in computeQpResidual()?
If you want to write a kernel that is valid for more than nodal variabless one reason.

**Reply by Ethan-xj** [reply]
But when I set the variable first lagrange and I add the code which can output the qp value of variable like：std<<cout<<_vel［_qp］<<endl； in kernel，it outputs some values indeed，I wanna know where does that “qp value” come from？
When I set 6 nodes in a 1D mesh and set the variable first lagrange，it outputs 10 values（I think FISRT stands that there are two qp point between two nodes，and plus the number of the element（6-1=5））.
I cant understand Orz

**Reply by GiudGiud** [reply]
6 nodes -> 5 elements 
If there's 2 qps per element, then that's how you got 10 values

The quadrature point value can be computed by evaluating the shape functions at the quadrature points.

**Reply by Ethan-xj** [reply]
So when I set a nodal variable, it actually is computed at each node, not quadrature point. The value of quadrature point is computed by evaluating, but can't decide the value of node. And my previous understanding is the way of elemental variable's calculation
, am I right?
So nodal variable: node value decides quadrature point value;
elemental variable: quadrature point value decides node value.
Can I simplify it this way?

**Reply by GiudGiud** [reply]
The values are stored at each node (they are stored in an array, mapped to nodes, we dont store value directly on the mesh objects)

The variable values are computed at quadrature points when we evaluate the residual or the Jacobian.

nodal variable: yes
elemental variable: no. The node value of an elemental variable may be undefined, because a node is shared by multiple elements and the variable is discontinuous for example
Also, the Qp value is not used to compute values. The Qp value is computed by evaluating the shape function(s). The dofs of an elemental variables may relate to arbitrary points on an element, determined by the finite element family and its order

This documentation should be helpful 
https://mooseframework.inl.gov/help/finite_element_concepts/index.html

**Reply by Ethan-xj** [reply]
Thank you a lot! I will check the documentation with your answer :D
